Add custom extension mapping

jswanner opened the pull request:
Thanks for the great library!

This PR is probably a bit odd, so I understand if you're not interested in accepting it.

Basically, I was wanting to use MJML for HTML email templates, and I couldn't figure out an elegant way to make it work and ended up with this.

It allows formats to be customized so that for other template extensions can be used. For instance, to make MJML work:

```elixir
defmodule Notifier do
  use Phoenix.Swoosh, formats: %{"html" => "mjml", "text" => "text"}
end

config :phoenix, format_encoders: [mjml: Phoenix.Mjml]

defmodule Phoenix.Mjml do
  def encode_to_iodata!(mjml) do
    with {:ok, html} <- Mjml.to_html(mjml) do
      html
    end
  end
end
```

If you're interested but want some changes, let me know. Thanks!

princemaple commented:
I will probably review this one late next week. Cheers.

princemaple commented:
1. https://github.com/phoenixframework/phoenix_template/blob/v1.0.1/lib/phoenix/template.ex Looks like the `format_encoders` have been moved into `phoenix_template` and can be missing in Phoenix 1.7+. Shall we target the latest version?
2. I don't think it's a valid use case to customize the format at runtime. Can we store the option in `use` to a module attribute, and use grab the attribute at runtime? I think customization per file is enough.
3. We have fixed target formats and dynamic custom extensions here, so we can probably reverse the lookup order to ext => format, and in render calls, we simply find out which format the file (the template) should render to, i.e. user calls `render_body(email, "a.custom", assigns)` and we just figure out whether it should go into `html_body` or `text_body`.

jswanner commented:
> 1. https://github.com/phoenixframework/phoenix_template/blob/v1.0.1/lib/phoenix/template.ex Looks like the `format_encoders` have been moved into `phoenix_template` and can be missing in Phoenix 1.7+. Shall we target the latest version?

Unless I'm missing something, I don't think this is something we need to worry about here. The example above where I configured `format_encoders`, that's what I did in my application after using this fork of Phoenix.Swoosh to hook everything up. So, I believe at least, this will be a documentation concern when it comes to `format_encoders` -- which I'm happy to help with. 

> 2. I don't think it's a valid use case to customize the format at runtime. Can we store the option in `use` to a module attribute, and use the attribute at runtime? I think customization per file is enough.

Fair enough. That was more or less my plan going in, but the way I implemented it seemed more inline with how views and layouts are handled. I'm happy to make this change. 

> 3. We have fixed target formats and dynamic custom extensions here, so we can probably reverse the lookup order to ext => format, and in render calls, we simply find out which format the file (the template) should render to, i.e. user calls `render_body(email, "a.custom", assigns)` and we just figure out whether it should go into `html_body` or `text_body`.

Okay, I can look into making this change as well. 

Thanks for your feedback. I should have some time to work on this at some point in the next few days.

princemaple commented:
🙏 appreciate the work. Thanks!

jswanner commented:
@princemaple, 

> 2\. I don't think it's a valid use case to customize the format at runtime. Can we store the option in `use` to a module attribute, and use the attribute at runtime? I think customization per file is enough.

This turned out to be less straight forward than I thought. Since the rendering is done in the `Phoenix.Swoosh` module (and not the module that `use Phoenix.Swoosh`), a module attribute didn't feel appropriate, since it would need to be an attribute on `Phoenix.Swoosh`. So, it seems to me at least, that we need to pass the customization as data to `Phoenix.Swoosh.render_body/3` somehow, either by changing the function's signature or stuffing it in `email.private`.

> 3\. We have fixed target formats and dynamic custom extensions here, so we can probably reverse the lookup order to ext => format, and in render calls, we simply find out which format the file (the template) should render to, i.e. user calls `render_body(email, "a.custom", assigns)` and we just figure out whether it should go into `html_body` or `text_body`.

In addition to reversing the order, I decided to remove the indirection of `"custom" -> "html" -> :html_body` and instead have the customization specify the "body key," as in:

```elixir
defmodule Notifier do
  use Phoenix.Swoosh, formats: %{"mjml" => :html_body, "text" => :text_body}
end
```

Let me know what you think, if there are further changes you'd like, and thanks for your time!

princemaple commented:
Thanks for the update and the explanation. I'll get back to you in a few days. Cheers, have a great week.

princemaple commented:
Hi @jswanner thanks for the patience. I made some tiny tweaks, please take a look when you have time. If all good, I'll merge this in.

jswanner commented:
@princemaple, looks good to me! Thanks for your work on this! ❤️

princemaple commented:
💜 

Would you like to add some docs? Or I will do it in a couple days and make a release.
